### PR TITLE
Add option to output matrix in elliptic solver

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Allows to build the explicit matrix representation of the linear operator column-by-column. This is useful for debugging and analysis. Here's how to do it (@PunJustice):

1. See the last commit in this PR for how to add `BuildMatrix` as an optional phase to the executable. It currently just builds the matrix once, but we can add a feature to build it in every step of an outer iteration if needed.
2. Read the matrix from the H5 files in Python like this (you can use `./bin/python-spectre -m jupyterlab` to quickly spin up a Jupyter server in which you can import the `spectre` Python modules. You can also[ connect to it in VSCode](https://code.visualstudio.com/docs/datascience/jupyter-notebooks#_connect-to-a-remote-jupyter-server).):

```py
import spectre.IO.H5 as spectre_h5
import matplotlib.pyplot as plt
from spectre.Elliptic.ReadH5 import read_matrix

with spectre_h5.H5File("Volume0.h5", "r") as h5file:
    matrix_file = h5file.get_vol("/Matrix")
    matrix = read_matrix(matrix_file)

plt.spy(matrix, precision=1e-10)
```

Output:

65b3d0a0-8176-4216-82b4-7925c4dfde35

(This is a 2D DG Laplace operator on 2x2 elements with 6x6 grid points each)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@wthrowe would you be willing to review this?

File: src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+  using type = std::string;
+  using group = BuildMatrixOptionsGroup;
+  static constexpr Options::String help = {
+      "Subfile name in the volume data H5 files where the matrix will be "
+      "stored. Each observation in the subfile is a column of the matrix. The "
+      "row index is the order of elements defined by the ElementId in the "
+      "volume data, by the order of tensor components encoded in the name of "
+      "the components, and by the contiguous ordering of grid points for each "
+      "component."};
+  using option_tags = tmpl::list<MatrixSubfileName>;
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp]: As far as I know, the discussion in #1345 has not been superseded and simple tags and option tags are supposed to be distinct and in different namespaces.

File: src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+  using type = size_t;
+};
+
+}  // namespace Tags
+
+namespace Actions {
+
+namespace detail {
+
+/// \brief The total number of grid points (size of the matrix) and the index of
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp:92]: Documentation doesn't show up in detail namespaces.  You've done a good job documenting this stuff, so maybe make it non-detail?

File: src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+    const auto [total_num_points, local_first_index] =
+        detail::total_num_points_and_local_first_index(
+            element_id, num_points_per_element,
+            OperandTag::type::number_of_independent_components);
+    if (get<logging::Tags::Verbosity<Tags::BuildMatrixOptionsGroup>>(box) >=
+            Verbosity::Quiet and
+        local_first_index == 0) {
+      Parallel::printf("Building explicit matrix representation of size " +
+                       std::to_string(total_num_points) + " x " +
+                       std::to_string(total_num_points) + ".\n");
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp]:
```c++
Parallel::printf(
    "Building explicit matrix representation of size %zu x %zu.\n",
    total_num_points, total_num_points);
```

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp (hunk 2 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/LinearSolver/Actions/BuildMatrix.hpp:92]: Not sure, these functions aren't intended to be part of the public interface, so no need to make them public IMO. I just wanted to document them and Doxygen syntax is pretty good for that, and consistent with other docs. I think it's fine to use Doxygen syntax even if detail things are filtered out. Makes it easy to make functions public if we decide to do so later.

**Comment by nilsvu** [comment]
Pushed a fixup. Thanks for reviewing @wthrowe !

**Review by wthrowe — APPROVED** [review]
Sounds good.  Squash.

**Pull request merged**